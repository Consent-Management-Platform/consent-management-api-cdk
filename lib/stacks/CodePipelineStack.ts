import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Effect, OpenIdConnectPrincipal, OpenIdConnectProvider, PolicyDocument, PolicyStatement, Role } from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

import { GITHUB_ORGANIZATION } from '../constants/github';
import { StageConfig } from '../interfaces/stage-config';

export interface CodePipelineStackProps extends StackProps {
  stageConfig: StageConfig;
}

/**
 * Defines infrastructure supporting automated deployment pipelines.
 */
export class CodePipelineStack extends Stack {
  constructor(scope: Construct, id: string, readonly props: CodePipelineStackProps) {
    super(scope, id, props);

    // Create GitHub OpenID Connect provider
    const oidcProvider = new OpenIdConnectProvider(this, 'GitHubOIDCProvider', {
      url: 'https://token.actions.githubusercontent.com',
      clientIds: ['sts.amazonaws.com'],
    });

    // Create GitHub principal which will be used to assume deployment roles.
    const gitHubPrincipal = new OpenIdConnectPrincipal(oidcProvider).withConditions(
      {
        StringLike: {
          'token.actions.githubusercontent.com:sub':
            `repo:${GITHUB_ORGANIZATION}/*:*`,
        },
      }
    );

    /**
      * Create a deployment role that has short lived credentials. The only
      * principal that can assume this role is the GitHub Open ID provider.
      *
      * This role is granted authority to assume AWS CDK roles.
      */
    new Role(this, 'GitHubActionsRole', {
      assumedBy: gitHubPrincipal,
      description: 'Role assumed by GitHub Actions to deploy code/infra using aws cdk',
      roleName: 'github-ci-role',
      maxSessionDuration: Duration.hours(1),
      inlinePolicies: {
        CdkDeploymentPolicy: new PolicyDocument({
          assignSids: true,
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['sts:AssumeRole'],
              resources: [`arn:aws:iam::${this.account}:role/cdk-*`],
            }),
          ],
        }),
      },
    });
  }
}
