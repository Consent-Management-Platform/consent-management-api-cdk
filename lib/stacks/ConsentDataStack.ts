import { Stack, StackProps } from 'aws-cdk-lib';
import { AttributeType, BillingMode, ProjectionType, Table, TableEncryption } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';

import { StageConfig } from '../interfaces/stage-config';

export interface ConsentDataStackProps extends StackProps {
  stageConfig: StageConfig;
}

/**
 * Defines the consent data storage layer.
 *
 * This stack is separated from higher-level components such as API compute services
 * to minimize the risk of other layer changes impacting the consent database,
 * which should be extremely stable and never deleted after prod launch.
 */
export class ConsentDataStack extends Stack {
  public readonly consentTable: Table;

  constructor(scope: Construct, id: string, readonly props: ConsentDataStackProps) {
    super(scope, id, props);

    this.consentTable = this.createConsentTable();
    this.createConsentsByServiceUserGsi(this.consentTable);
  }

  private createConsentTable() {
    return new Table(this, 'ServiceUserConsentDynamoDBTable', {
      tableName: 'ServiceUserConsent',
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: {
        name: 'Id',
        type: AttributeType.STRING
      },
      pointInTimeRecovery: true,
      deletionProtection: true,
      billingMode: BillingMode.PAY_PER_REQUEST
    });
  }

  private createConsentsByServiceUserGsi(consentTable: Table) {
    consentTable.addGlobalSecondaryIndex({
      indexName: 'ConsentsByServiceUser',
      partitionKey: {
        name: 'UserId',
        type: AttributeType.STRING
      },
      sortKey: {
        name: 'ServiceId',
        type: AttributeType.STRING
      },
      projectionType: ProjectionType.ALL
    });
  }
}
