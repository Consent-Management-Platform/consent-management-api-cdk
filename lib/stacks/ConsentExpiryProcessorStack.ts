import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Table } from 'aws-cdk-lib/aws-dynamodb';
import { PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Code, Function, Runtime, SnapStartConf } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

import { StageConfig } from '../interfaces/stage-config';
import { CustomLambdaFunction } from '../constructs/CustomLambdaFunction';
import { Rule, Schedule } from 'aws-cdk-lib/aws-events';
import { LambdaFunction } from 'aws-cdk-lib/aws-events-targets';

export interface ConsentExpiryProcessorStackProps extends StackProps {
  codePackageFilePath: string;
  consentTable: Table;
  stageConfig: StageConfig;
}

/**
 * Defines infrastructure for automatically updating consents to EXPIRED when they have passed their expiry time.
 */
export class ConsentExpiryProcessorStack extends Stack {
  public readonly expiryProcessorLambda: Function;

  constructor(scope: Construct, id: string, readonly props: ConsentExpiryProcessorStackProps) {
    super(scope, id, props);

    this.expiryProcessorLambda = this.createExpiryProcessorLambdaFunction();
    this.createLambdaSchedule(this.expiryProcessorLambda);
  }

  private createExpiryProcessorLambdaFunction(): Function {
    const lambdaFunction: Function = new CustomLambdaFunction(this, 'ConsentExpiryProcessorLambda', {
      code: Code.fromAsset(this.props.codePackageFilePath),
      description: 'Consent Expiry Processor Lambda',
      handler: 'com.consentframework.consentexpiryprocessor.ConsentExpiryProcessor::handleRequest',
      memorySize: 1536,
      runtime: Runtime.JAVA_21,
      snapStart: SnapStartConf.ON_PUBLISHED_VERSIONS,
      timeout: Duration.minutes(15)
    });

    // Grant permissions to query DynamoDB consent data
    lambdaFunction.addToRolePolicy(new PolicyStatement({
      sid: 'ConsentDynamoDBQueryPermissions',
      actions: [
        'dynamodb:ConditionCheckItem',
        'dynamodb:Query',
        'dynamodb:Scan',
        'dynamodb:UpdateItem'
      ],
      resources: [
        this.props.consentTable.tableArn,
        `${this.props.consentTable.tableArn}/index/*`
      ]
    }));

    return lambdaFunction;
  }

  // Schedule the Lambda function to run hourly
  private createLambdaSchedule(apiLambda: Function) {
    // Choose a random minute and distribute with any other scheduled jobs to avoid load spikes
    const scheduleRule = new Rule(this, 'ConsentExpiryProcessorSchedule', {
      schedule: Schedule.cron({
        minute: '51'
      })
    });

    scheduleRule.addTarget(new LambdaFunction(apiLambda));
  }
}
