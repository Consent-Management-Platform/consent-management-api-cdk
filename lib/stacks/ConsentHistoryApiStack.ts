import { Aws, CfnOutput, Duration, Stack, StackProps } from 'aws-cdk-lib';
import { ApiDefinition, Deployment, EndpointType, MethodLoggingLevel, SpecRestApi } from 'aws-cdk-lib/aws-apigateway';
import { Table } from 'aws-cdk-lib/aws-dynamodb';
import { AccountRootPrincipal, ArnPrincipal, Effect, PolicyDocument, PolicyStatement, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { Code, Function, Runtime, SnapStartConf } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';
import { join } from 'path';

import {
  CONSENT_HISTORY_API_DOCS_URL,
  CONSENT_HISTORY_API_ENDPOINT_EXPORT_NAME,
  CONSENT_HISTORY_API_GATEWAY_ID_EXPORT_NAME,
  CONSENT_HISTORY_API_THROTTLING_BURST_LIMIT,
  CONSENT_HISTORY_API_THROTTLING_LIMIT
} from '../constants/apis';
import { StageConfig } from '../interfaces/stage-config';
import { constructApiDefinition } from '../utils/openapi';
import { CustomLambdaFunction } from '../constructs/CustomLambdaFunction';

export interface ConsentHistoryApiStackProps extends StackProps {
  apiCodePackageFilePath: string;
  codeDeployRole: Role;
  consentHistoryTable: Table;
  stageConfig: StageConfig;
}

/**
 * Defines Consent History API gateway and compute infrastructure.
 */
export class ConsentHistoryApiStack extends Stack {
  public readonly apiLambda: Function;
  public readonly restApi: SpecRestApi;

  constructor(scope: Construct, id: string, readonly props: ConsentHistoryApiStackProps) {
    super(scope, id, props);

    this.apiLambda = this.createApiLambdaFunction();
    this.restApi = this.createRestApiGateway(this.apiLambda);
    this.grantCodeDeployRolePermissions(this.restApi);

    this.createExports(this.restApi);
  }

  private createApiLambdaFunction(): Function {
    const lambdaFunction: Function = new CustomLambdaFunction(this, 'ConsentHistoryApiLambda', {
      code: Code.fromAsset(this.props.apiCodePackageFilePath),
      description: 'Consent History API Lambda',
      handler: 'com.consentframework.consenthistory.api.ConsentHistoryApiService::handleRequest',
      memorySize: 1536,
      runtime: Runtime.JAVA_21,
      snapStart: SnapStartConf.ON_PUBLISHED_VERSIONS,
      timeout: Duration.minutes(1)
    });

    // Grant permissions to query DynamoDB consent history data
    lambdaFunction.addToRolePolicy(new PolicyStatement({
      sid: 'ConsentHistoryDdbQueryPermissions',
      actions: [
        'dynamodb:ConditionCheckItem',
        'dynamodb:GetItem',
        'dynamodb:Query'
      ],
      resources: [
        this.props.consentHistoryTable.tableArn,
        `${this.props.consentHistoryTable.tableArn}/index/*`
      ]
    }));

    return lambdaFunction;
  }

  private createRestApiGateway(apiLambda: Function): SpecRestApi {
    const openApiSpecFilePath = join(__dirname, '../../resources/ConsentHistoryApi.openapi.json');
    const apiDefinition: object = constructApiDefinition(this.props.env!, apiLambda.functionArn, openApiSpecFilePath);
    const apiAccessPolicy = this.createApiGatewayAccessPolicy();

    const api: SpecRestApi = new SpecRestApi(this, 'Consent History API Gateway', {
      apiDefinition: ApiDefinition.fromInline(apiDefinition),
      cloudWatchRole: true,
      deployOptions: {
        metricsEnabled: true,
        loggingLevel: MethodLoggingLevel.INFO,
        stageName: this.props.stageConfig.stage,
        throttlingBurstLimit: CONSENT_HISTORY_API_THROTTLING_BURST_LIMIT,
        throttlingRateLimit: CONSENT_HISTORY_API_THROTTLING_LIMIT
      },
      description: `Consent History API, see documentation at ${CONSENT_HISTORY_API_DOCS_URL}`,
      endpointExportName: CONSENT_HISTORY_API_ENDPOINT_EXPORT_NAME,
      endpointTypes: [EndpointType.EDGE],
      policy: apiAccessPolicy,
      restApiName: 'ConsentHistoryApi'
    });

    // Trigger API Gateway deployment when there are OpenAPI spec updates to apply the changes
    const apiDeployment = new Deployment(this, 'ConsentHistoryApiDeployment', { api });
    apiDeployment.addToLogicalId(apiDefinition);

    apiLambda.grantInvoke(new ServicePrincipal('apigateway.amazonaws.com').withConditions({
      ArnLike: { 'aws:SourceArn': api.arnForExecuteApi() }
    }));

    return api;
  }

  /**
   * Explicitly set which principals are allowed to access which API methods.
   * Anyone not explicitly granted permissions will be denied API access.
   *
   * Ref: https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-authorization-flow.html
   */
  private createApiGatewayAccessPolicy(): PolicyDocument {
    const githubCiRolePrincipal = new ArnPrincipal(`arn:aws:iam::${Aws.ACCOUNT_ID}:role/github-ci-role`);

    return new PolicyDocument({
      statements: [
        new PolicyStatement({
          sid: 'ConsentHistoryApiGatewayFullInvokeAccess',
          actions: ['execute-api:Invoke'],
          effect: Effect.ALLOW,
          principals: [
            // Grant services within the same account API access
            new AccountRootPrincipal(),
            // Grant the CD/CI role API access
            githubCiRolePrincipal,
          ],
          // Grants invoke permissions to all APIs within this API Gateway
          resources: ['execute-api/*']
        }),
        new PolicyStatement({
          sid: 'ConsentHistoryApiGatewayApiAccess',
          actions: [
            'apigateway:GET',
            'apigateway:POST'
          ],
          effect: Effect.ALLOW,
          principals: [
            githubCiRolePrincipal,
          ],
          resources: [
            `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}::/restapis/*/resources`,
            `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}::/restapis/*/resources/*/*/*`,
          ]
        })
      ]
    });
  }

  private grantCodeDeployRolePermissions(restApi: SpecRestApi): void {
    // Allow to invoke Consent History API endpoints for API integ tests
    this.props.codeDeployRole.addToPrincipalPolicy(new PolicyStatement({
      sid: 'ConsentHistoryApiInvokePermissions',
      actions: ['execute-api:Invoke'],
      effect: Effect.ALLOW,
      resources: [restApi.arnForExecuteApi()],
    }));
    // Allow to invoke API endpoints via API Gateway TestInvokeApi for AWS CLI based integ tests
    this.props.codeDeployRole.addToPrincipalPolicy(new PolicyStatement({
      sid: 'ConsentHistoryApiTestInvokeMethodPermissions',
      actions: [
        'apigateway:GET',
        'apigateway:POST'
      ],
      effect: Effect.ALLOW,
      resources: [
        `arn:aws:apigateway:us-west-2::/restapis/${restApi.restApiId}/resources`,
        `arn:aws:apigateway:us-west-2::/restapis/${restApi.restApiId}/resources/*/*/*`
      ],
    }));

    // Allow to generate API clients from API Gateway endpoint
    // These encapsulate the logic for making authenticated queries, building requests, and parsing API responses
    this.props.codeDeployRole.addToPrincipalPolicy(new PolicyStatement({
      sid: 'ConsentHistoryApiGetSdkPermissions',
      actions: ['apigateway:GET'],
      effect: Effect.ALLOW,
      resources: [`arn:${Aws.PARTITION}:apigateway:${Aws.REGION}::/restapis/${restApi.restApiId}/stages/*/sdks/*`],
    }));

    // Allow to list CDK exports to retrieve API Gateway ID
    this.props.codeDeployRole.addToPrincipalPolicy(new PolicyStatement({
      sid: 'CloudFormationListExportsPermissions',
      actions: ['cloudformation:ListExports'],
      effect: Effect.ALLOW,
      resources: ['*'], // CloudFormation exports are global, cannot filter permissions
    }));
  }

  private createExports(restApi: SpecRestApi): void {
    new CfnOutput(this, 'ConsentHistoryApiGatewayIdExport', {
      exportName: CONSENT_HISTORY_API_GATEWAY_ID_EXPORT_NAME,
      value: restApi.restApiId
    });
  }
}
