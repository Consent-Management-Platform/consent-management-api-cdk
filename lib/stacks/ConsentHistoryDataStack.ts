import { Stack, StackProps } from 'aws-cdk-lib';
import { AttributeType, BillingMode, Table, TableEncryption } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';

import { StageConfig } from '../interfaces/stage-config';

export interface ConsentHistoryDataStackProps extends StackProps {
  stageConfig: StageConfig;
}

/**
 * Defines the consent history data storage layer.
 */
export class ConsentHistoryDataStack extends Stack {
  public readonly consentHistoryTable: Table;

  constructor(scope: Construct, id: string, readonly props: ConsentHistoryDataStackProps) {
    super(scope, id, props);

    this.consentHistoryTable = this.createConsentHistoryTable();
  }

  private createConsentHistoryTable() {
    return new Table(this, 'ServiceUserConsentHistoryDynamoDBTable', {
      tableName: 'ConsentHistory',
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: {
        name: 'id',
        type: AttributeType.STRING
      },
      sortKey: {
        name: 'eventId',
        type: AttributeType.STRING
      },
      pointInTimeRecovery: true,
      deletionProtection: true,
      billingMode: BillingMode.PAY_PER_REQUEST
    });
  }
}
