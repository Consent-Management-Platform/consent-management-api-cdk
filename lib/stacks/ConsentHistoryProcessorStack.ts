import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Table } from 'aws-cdk-lib/aws-dynamodb';
import { Code, Function, Runtime, SnapStartConf, StartingPosition } from 'aws-cdk-lib/aws-lambda';
import { DynamoEventSource, SqsDlq } from 'aws-cdk-lib/aws-lambda-event-sources';
import { Queue, QueueEncryption } from 'aws-cdk-lib/aws-sqs';
import { Construct } from 'constructs';

import { CustomLambdaFunction } from '../constructs/CustomLambdaFunction';
import { StageConfig } from '../interfaces/stage-config';

export interface ConsentHistoryProcessorStackProps extends StackProps {
  codePackageFilePath: string;
  consentTable: Table;
  consentHistoryTable: Table;
  stageConfig: StageConfig;
}

/**
 * Defines the infrastructure for syncing consent updates to the consent history table.
 */
export class ConsentHistoryProcessorStack extends Stack {
  constructor(scope: Construct, id: string, readonly props: ConsentHistoryProcessorStackProps) {
    super(scope, id, props);
    this.createConsentHistoryProcessorLambda();
  }

  private createConsentHistoryProcessorLambda() {
    const lambdaFunction: Function = new CustomLambdaFunction(this, 'ConsentHistoryProcessorLambda', {
      code: Code.fromAsset(this.props.codePackageFilePath),
      description: 'Consent History Processor Lambda',
      environment: {
        CONSENT_HISTORY_TABLE_NAME: this.props.consentHistoryTable.tableName,
        STAGE: this.props.stageConfig.stage
      },
      handler: 'com.consentframework.consenthistory.consentingestor.ConsentStreamIngestor::handleRequest',
      memorySize: 1024,
      runtime: Runtime.JAVA_21,
      snapStart: SnapStartConf.ON_PUBLISHED_VERSIONS,
      timeout: Duration.minutes(1)
    });

    // Create a dead letter queue to collect events that failed processing
    const deadLetterQueue: Queue = new Queue(this, 'ConsentHistoryProcessorLambdaDlq', {
      queueName: 'ConsentHistoryProcessorLambdaDlq',
      encryption: QueueEncryption.SQS_MANAGED,
      retentionPeriod: Duration.days(14)
    });

    // Add consent table's DynamoDB Stream as an event source for the Lambda function
    this.props.consentTable.grantStreamRead(lambdaFunction);
    lambdaFunction.addEventSource(new DynamoEventSource(this.props.consentTable, {
      startingPosition: StartingPosition.TRIM_HORIZON,
      onFailure: new SqsDlq(deadLetterQueue)
    }));

    // Grant the Lambda function permissions to write to the consent history table
    this.props.consentHistoryTable.grantWriteData(lambdaFunction);
  }
}
