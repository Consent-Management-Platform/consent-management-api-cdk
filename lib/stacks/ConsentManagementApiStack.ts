import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { ApiDefinition, EndpointType, MethodLoggingLevel, SpecRestApi } from 'aws-cdk-lib/aws-apigateway';
import { AccountRootPrincipal, Effect, PolicyDocument, PolicyStatement, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { Code, Function, Runtime } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';
import { readFileSync } from 'fs';
import { join } from 'path';

import {
  CONSENT_MANAGEMENT_API_DOCS_URL,
  CONSENT_MANAGEMENT_API_ENDPOINT_EXPORT_NAME,
  CONSENT_MANAGEMENT_API_THROTTLING_BURST_LIMIT,
  CONSENT_MANAGEMENT_API_THROTTLING_LIMIT
} from '../constants/apis';
import { StageConfig } from '../interfaces/stage-config';

export interface ConsentManagementApiStackProps extends StackProps {
  apiCodePackageFilePath: string;
  stageConfig: StageConfig;
}

/**
 * Defines Consent Management API gateway and compute infrastructure.
 */
export class ConsentManagementApiStack extends Stack {
  public readonly apiLambda: Function;
  public readonly restApi: SpecRestApi;

  constructor(scope: Construct, id: string, readonly props: ConsentManagementApiStackProps) {
    super(scope, id, props);

    this.apiLambda = this.createApiLambdaFunction();
    this.restApi = this.createRestApiGateway(this.apiLambda);
  }

  private createApiLambdaFunction(): Function {
    return new Function(this, 'Consent Management API Lambda', {
      code: Code.fromAsset(this.props.apiCodePackageFilePath),
      description: 'Consent Management API Lambda',
      handler: 'com.consentframework.consentmanagement.api.ConsentManagementApiService::handleRequest',
      memorySize: 512,
      runtime: Runtime.JAVA_21,
      timeout: Duration.minutes(1)
    });
  }

  private createRestApiGateway(apiLambda: Function): SpecRestApi {
    const apiDefinition = this.constructApiDefinition(apiLambda);
    const apiAccessPolicy = this.createApiGatewayAccessPolicy();

    const api: SpecRestApi = new SpecRestApi(this, 'Consent Management API Gateway', {
      apiDefinition: ApiDefinition.fromInline(apiDefinition),
      cloudWatchRole: true,
      deployOptions: {
        metricsEnabled: true,
        loggingLevel: MethodLoggingLevel.INFO,
        stageName: 'dev',
        throttlingBurstLimit: CONSENT_MANAGEMENT_API_THROTTLING_BURST_LIMIT,
        throttlingRateLimit: CONSENT_MANAGEMENT_API_THROTTLING_LIMIT
      },
      description: `Consent Management API, see documentation at ${CONSENT_MANAGEMENT_API_DOCS_URL}`,
      endpointExportName: CONSENT_MANAGEMENT_API_ENDPOINT_EXPORT_NAME,
      endpointTypes: [EndpointType.EDGE],
      policy: apiAccessPolicy,
      restApiName: 'ConsentManagementApi'
    });

    apiLambda.grantInvoke(new ServicePrincipal('apigateway.amazonaws.com').withConditions({
      ArnLike: { 'aws:SourceArn': api.arnForExecuteApi() }
    }));

    return api;
  }

  private constructApiDefinition(apiLambda: Function) {
    const openApiSpecFilePath = join(__dirname, '../../resources/ConsentManagementApi.openapi.json');
    const apiDefinition = JSON.parse(readFileSync(openApiSpecFilePath, 'utf8'));

    // Add API Gateway integration to OpenAPI spec
    for (const path in apiDefinition.paths) {
      for (const operation in apiDefinition.paths[path]) {
        apiDefinition.paths[path][operation]['x-amazon-apigateway-integration'] = {
          type: 'aws_proxy',
          httpMethod: 'POST',
          passthroughBehavior: 'when_no_match',
          responses: {
            default: {
              statusCode: '200',
            },
          },
          uri: `arn:aws:apigateway:${this.props.env!.region}:lambda:path/2015-03-31/functions/${apiLambda.functionArn}/invocations`
        }
      }
    }

    return apiDefinition;
  }

  /**
   * Explicitly set which principals are allowed to access which API methods.
   * Anyone not explicitly granted permissions will be denied API access.
   *
   * Ref: https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-authorization-flow.html
   */
  private createApiGatewayAccessPolicy(): PolicyDocument {
    return new PolicyDocument({
      statements: [new PolicyStatement({
        sid: 'ConsentManagementApiGatewaySameAccountFullInvokeAccess',
        actions: ['execute-api:Invoke'],
        effect: Effect.ALLOW,
        // Grant services within the same account API access
        principals: [new AccountRootPrincipal()],
        // Grants invoke permissions to all APIs within this API Gateway
        resources: ['execute-api/*']
      })]
    });
  }
}
