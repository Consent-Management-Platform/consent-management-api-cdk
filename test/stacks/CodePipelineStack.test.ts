import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { MOCK_ENV } from '../../fixtures/mock-env';
import { MOCK_STAGE_CONFIG } from '../../fixtures/mock-stage-config';
import { CodePipelineStack } from '../../lib/stacks/CodePipelineStack';

describe('CodePipelineStack', () => {
  it('creates the expected CloudFormation template from CDK', () => {
    const app = new App();
    const stack = new CodePipelineStack(app, 'CodePipelineStack', {
      env: MOCK_ENV,
      stageConfig: MOCK_STAGE_CONFIG
    });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  })
});
