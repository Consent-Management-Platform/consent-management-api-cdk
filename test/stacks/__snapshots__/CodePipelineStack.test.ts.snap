// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodePipelineStack creates the expected CloudFormation template from CDK 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0": {
      "DependsOn": [
        "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-2",
          "S3Key": "62fa02efcaa700e1c247e1d3cc2aa0cd07a0808a9a3e3d2230e51f57a02233fb.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "iam:CreateOpenIDConnectProvider",
                    "iam:DeleteOpenIDConnectProvider",
                    "iam:UpdateOpenIDConnectProviderThumbprint",
                    "iam:AddClientIDToOpenIDConnectProvider",
                    "iam:RemoveClientIDFromOpenIDConnectProvider",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GitHubActionsRole4F1BBA26": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "StringLike": {
                  "token.actions.githubusercontent.com:sub": "repo:Consent-Management-Platform/*:*",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": {
                  "Ref": "GitHubOIDCProvider4F328116",
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Role assumed by GitHub Actions to deploy code/infra using aws cdk",
        "MaxSessionDuration": 3600,
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "sts:AssumeRole",
                  "Effect": "Allow",
                  "Resource": "arn:aws:iam::123456789012:role/cdk-*",
                  "Sid": "0",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "CdkDeploymentPolicy",
          },
        ],
        "RoleName": "github-ci-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "GitHubOIDCProvider4F328116": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ClientIDList": [
          "sts.amazonaws.com",
        ],
        "CodeHash": "62fa02efcaa700e1c247e1d3cc2aa0cd07a0808a9a3e3d2230e51f57a02233fb",
        "RejectUnauthorized": false,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0",
            "Arn",
          ],
        },
        "Url": "https://token.actions.githubusercontent.com",
      },
      "Type": "Custom::AWSCDKOpenIdConnectProvider",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
